discoverd/main.go

```diff
@@ -246,10 +283,124 @@ func waitHostDNSConfig() (addr string, resolvers []string) {
+	// error out if the store is loaded
+	if m.store != nil {
+		return fmt.Errorf("can't join, already a raft peer")
+	}
+	// error out of the handler isn't running in proxy mode
+	if !m.handler.Proxy.Load().(bool) {
+		return fmt.Errorf("can't promote, handler not in proxy mode")
+	}
+	// TODO(jpg): Replace this with better way of getting leader
+	randAddr := m.peers[rand.Intn(len(m.peers))]
+	m.logger.Println("getting leader using peer", randAddr)
+	client := discoverd.NewClientWithURL(randAddr)
+	leader, err := client.RaftLeader()
+	if err != nil {
+		m.logger.Println("failed to get leader addr from", randAddr)
+		return err
+	}
+	m.logger.Println("leader:", leader.Host)
+	// request the leader joins us to the cluster
+	m.logger.Println("requesting leader join us to cluster")
+	discoverd.NewClientWithURL(leader.Host).RaftAddPeer(m.advertiseAddr)
+	// open the store
+	m.logger.Println("opening store")
+	if err := m.openStore(true); err != nil {
+		return err
+	}
+	// TODO(jpg): Need updated peer list
+	//if !inPeerList(m.peers, m.advertiseAddr) {
+	//	m.peers = append(m.peers, m.advertiseAddr)
+	//}
+	//m.logger.Println("setting peers")
+	//m.store.SetPeers(m.peers)
+
+	// wait for our local store to catch up
+	// update the dns server to use the local store
+	m.logger.Println("switching dns server to real store")
+	m.dnsServer.SetStore(m.store)
+	// update the http server to use the local store
+	m.logger.Println("switching handler to real store")
+	m.handler.Store = m.store
+	m.handler.Proxy.Store(false)
+	// TODO(jpg): Update instance meta to mark outselves as a consensus peer
+	m.logger.Println("promoted")
+	return nil
+}
+
+// Leave the consensus set, demoting ourselves to proxy from raft node
+func (m *Main) Demote() error {
+	m.logger.Println("demotion requested")
+	// error out of we are not currently a raft peer
+	if m.store == nil {
+		return fmt.Errorf("can't leave, not currently a raft peer")
+	}
+	// error out of the handler is running in proxy mode
+	if m.handler.Proxy.Load().(bool) {
+		return fmt.Errorf("can't demote, handler in proxy mode")
+	}
+	// update the dns server to use proxy store
+	m.logger.Println("setting dns server to proxy store")
+	m.dnsServer.SetStore(&server.ProxyStore{Peers: m.peers})
+	// TODO(jpg): wait for http server to drain existing requests?
+	// update http server to use proxy store
+	m.logger.Println("setting handler to proxy mode for new requests")
+	m.handler.Proxy.Store(true)
+	// get the leader from the store
+	leaderAddr := m.store.Leader()
+	m.logger.Println("leader:", leaderAddr)
+
+	// we are the leader, run alternate routine
+	if leaderAddr == m.advertiseAddr {
+		m.store.RemovePeer(m.advertiseAddr)
+	} else {
+
+		// ask the leader to remove us from the peer set
+		m.logger.Println("asking the leader to remove us from the raft set")
+		discoverd.NewClientWithURL(leaderAddr).RaftRemovePeer(m.advertiseAddr)
+		//m.store.SetPeers(nil)
+	}
+	time.Sleep(3 * time.Second)
```

Comment:
Should have been a TODO comment here to look into waiting until our local peer store has applied the removePeer command from the leader. I haven't quite worked that out yet. It's maybe not super required because as long as it's been applied by the leader there should be no adverse effects of us simply shutting down the store.